[Security Solution] Improve rule import endpoint by splitting the work to multiple macrotasks

maximpn opened the pull request:
**Relates to:** https://github.com/elastic/kibana/issues/176207

## Summary

This PR sets increased socket timeout for rules import API endpoint `/api/detection_engine/rules/_import` to avoid situations when connection is interrupted while rules import is still in progress.

## Details

The original problem `Rule import creates extra rules while importing a large number of rules` is just a side effect of specific underlying behaviour caused by node.js. It has socket (TCP connection) timeout equals to 2 minutes by default. When the API endpoint takes longer than 2 minutes to be processed node.js terminates the TCP connection.

According to the [HTTP spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.4) the browser should retry if the connection was closed by the server. In this case it leads to a duplicate request importing exactly the same rules concurrently. In a presence if this race condition more than one rule having the same `rule_id` can be created.

The following has been done in this PR to mitigate the problem

- `timeout.idleSocket` is set to a 30 minutes
- each rules chunk import is wrapper into a macrotask via `setImmediate` to prevent event loop blocking
- `importRules` function has been refactored to improve readability

This PR doesn't address an issues caused by importing the same rules concurrently by different users. It also doesn't address handler early termination in case of external reasons like connection termination.


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

maximpn committed:
increase rule import's API endpoint socket timeout

maximpn committed:
simplify importRules helper function

maximpn committed:
add a unit test for the custom socket timeout

maximpn commented:
/ci

maximpn committed:
wrap rule import in macrotask to prevent event loop blocking

maximpn commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/utils/wrap_in_macrotask.ts:
```diff
@@ -0,0 +1,44 @@
+ * @returns a wrapped function
+ */
+export function wrapInMacrotask<Args extends unknown[], R>(
+  actionFn: (...args: Args) => Promise<R>
```
Does `actionFn` have to be a function returning a Promise? Or we can apply the approach to any functions?

maximpn replied:
I remove a restriction that `actionFn` have to be a function returning a Promise. It should help to use `wrapInMacrotask` for sync functions as well. While `setImmediate` can be used directly `wrapInMacrotask` wraps it in a Promise making it simpler to consume the result via `await`.

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/utils/wrap_in_macrotask.ts:
```diff
@@ -0,0 +1,44 @@
+ * @param actionFn function returning a promise
+ * @returns a wrapped function
+ */
+export function wrapInMacrotask<Args extends unknown[], R>(
```
It would be nice to have unit tests for the behavior of functions returned from `wrapInMacrotask`.

maximpn replied:
Sure, I've added unit tests.

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/utils/wrap_in_macrotask.ts:
```diff
@@ -0,0 +1,44 @@
+ * 2.0.
+ */
+
+/**
+ * Wraps an async function into a macro task to prevent blocking event loop.
+ *
+ * Node.js is built upon an event loop and has definition of micro and macro tasks.
+ *
+ * Macro tasks are functions scheduled via setTimeout, setImmediate and etc to be executed in event loop.
+ * Macro tasks scheduled in the current event loop cycle will be executed in the next cycle giving a chance to the other
+ * macro tasks to be executed.
+ *
+ * Micro tasks are also functions by scheduled via a Promise. Micro tasks get added to a micro task queue which must be
+ * depleted before event loop can move to the next cycle. It means that any microtask scheduled in the current event loop
+ * cycle will be executed in the same cycle. Potentially it can lead to blocking the event loop for longer periods or even
+ * infinitely.
+ *
+ * For more details check the following sources
+ * - [MDN Microtask guide](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide)
+ * - [MDN Microtask guide in_depth](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide/In_depth)
+ * - [What the heck is the event loop anyway? | Philip Roberts | JSConf EU](https://www.youtube.com/watch?v=8aGhZQkoFbQ)
+ *
+ *
+ * @param actionFn function returning a promise
+ * @returns a wrapped function
+ */
```
Thank you for this elaborate comment! Just a nit: please format it for readability so that each line is 80 or 100 characters at most.

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts:
```diff
@@ -46,116 +50,111 @@ export interface RuleExceptionsPromiseFromStreams {
+  };
+
+  for (const rulesChunk of ruleChunks) {
+    const chunkImportRuleResponses = await Promise.all(rulesChunk.map(wrapInMacrotask(importRule)));
```
Just an idea: should we run each chunk in a macrotask instead of each rule?

Alternative idea: should we use promise pool instead of chunking approach here?

maximpn replied:
I ended up wrapping rules chunk import in a macro task. Promise pool doesn't allow to do that without modification of its implementation.

banderror reviewed (commented):
Awesome PR @maximpn! Left a few questions and suggestions.

maximpn commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/api/rules/import_rules/route.ts:
```diff
@@ -51,6 +48,11 @@ export const importRulesRoute = (
           maxBytes: config.maxRuleImportPayloadBytes,
           output: 'stream',
         },
+        timeout: {
+          // Rule import rule may take a significant amount of time
+          // socket shouldn't be closed during this period
+          idleSocket: 30 * 60 * 1000,
+        },
```
I addressed this issue here https://github.com/elastic/kibana/pull/177329. Though I used 30 minutes timeout it's not a problem to change it to 1 hour. My plan is to update either this or the other [PR](https://github.com/elastic/kibana/pull/177329) depending on which one is merged first.

banderror reviewed (changes requested):
Tested locally using the provided ndjson containing 9930 rules. Can't confirm the fix.

I got a 408 request timeout after 18 minutes of handling the import request:

<img width="956" alt="Screenshot 2024-02-21 at 15 31 29" src="https://github.com/elastic/kibana/assets/7359339/3a194312-d396-4312-8c55-f4e7bbdb9b19">

Response:

```json
{"statusCode":408,"error":"Request Time-out","message":"Request Time-out"}
```

By the moment of receiving the 408 response it imported 9774 rules. However, on the server-side the handler continued to run, obviously, and in a few minutes after that it was already 14559 rules. There was no duplicate import calls, only one.

Something is still broken.

On top of that, I have big concerns around the performance in absolute numbers:

- I think an import of ~10k rules should take less than 18 minutes. Maybe a couple of minutes would make sense from the UX perspective.
- All the time while the import endpoint was executing, my local Kibana was severely overloaded to the extent that simple API calls happening in the background would take 15-30 seconds to run. There were messages in the logs about blocked event loop. I'd expect rule import to be a mostly I/O-heavy endpoint, so I don't get why the event loop was blocked.

It feels like something is wrong with the fix which affected the performance, or something was already wrong before the fix.

@maximpn Could you please:

- Deploy a BEFORE and an AFTER versions to Cloud using `ci:cloud-deploy`, so we could compare the performance in absolute numbers in similar non-local environments. Kibana ala carte could be used as well, I guess.
- Profile the performance before and after the fix locally using APM. You will need to have your personal Elastic deployment in production ESS and connect to it from local Kibana like that:

```yaml
elastic.apm.active: true
elastic.apm.serverUrl: https://456745674567.apm.europe-west3.gcp.cloud.es.io:443
elastic.apm.secretToken: sdfgaergesrgerg
elastic.apm.sampleRate: 1
```

We can chat about all that over zoom, of course, feel free to ping me.

maximpn committed:
increase socket timeout to 1 hour

maximpn committed:
move importRule helper function to a separate file

maximpn committed:
wrap comments

maximpn committed:
add unit tests for wrapInMacrotask

maximpn commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts:
```diff
@@ -46,116 +50,111 @@ export interface RuleExceptionsPromiseFromStreams {
-          }
+  const importRuleResponses = [...rulesResponseAcc];
+  const ruleChunks = chunk(IMPORT_RULES_CHUNK_SIZE, rules);
+  const importRule = async (ruleToImport: PromiseFromStreams): Promise<ImportRuleResponse> => {
```
sure

maximpn commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts:
```diff
@@ -46,116 +50,111 @@ export interface RuleExceptionsPromiseFromStreams {
+  };
+
+  for (const rulesChunk of ruleChunks) {
+    const chunkImportRuleResponses = await Promise.all(rulesChunk.map(wrapInMacrotask(importRule)));
```
I refactored this section in a different way

maximpn commented:
/ci

maximpn committed:
update a unit test

maximpn commented:
/ci

maximpn commented:
/ci

maximpn commented:
@banderror thank you for thorough testing 👍 

Yes, the bug is fixed only for Kibana running without a proxy. In fact we have Kibana running with a proxy almost all the time which means this PR doesn't fix anything.

I suggest to classify this PR as rule import refactoring for better performance (APM) visibility. To make it transparent I updated [[Security Solution] Set socket timeout for potentially long running Rule Management API endpoints #177329](https://github.com/elastic/kibana/pull/177329) by adding 1 hour socket timeout for `/api/detection_engine/rules/_import` API endpoint. On top of that I updated current PR by adding APM tracking of `importRulesChunk` and `importRule` functions performance.

[2024-03-06T12:11:58.015+01:00][WARN ][http.server.Kibana] Event loop utilization for /api/detection_engine/rules/_import exceeded threshold of 250ms (314475ms out of 2066006ms) and 15% (15%) 

I made thorough performance testing by comparing this PR with [[Security Solution] Set socket timeout for potentially long running Rule Management API endpoints #177329](https://github.com/elastic/kibana/pull/177329). The latter has just socket timeout increased. The major part of comparisons were made locally without a proxy. To run Kibana locally without a proxy `server.basePath` shouldn't be set in `kibana.yaml/`kibana.dev.yaml` and Kibana is run with `--no-base-path` param.

I've noticed similar results despite the change made here. It's not a surprise as this PR doesn't make any performance improvements. The only difference is in `wrapInMacrotask` to reduce event loop utilisation. I experimented with different rule chunk size (rules to import are split in chunks and imported in parallel). It turned out rule chunk size has significant impact on importing duration. I got the following results

| Chunk size | w/ `wrapInMacrotask` | w/o `wrapInMacrotask` | Event loop utilization warnings w/ `wrapInMacrotask` | Event loop utilization warnings w/o `wrapInMacrotask` |
| ----------- | ------------------------ | --------------------------- | ------------------------------- |
|       50         |   ~3 min 29 sec                |     ~3 min 30 sec                  |  Event loop utilization for /api/detection_engine/rules/_import exceeded threshold of 250ms (140495ms out of 206494ms) and 15% (68%) 
 | exceeded threshold of 250ms (136334ms out of 209992ms) and 15% (65%) | 
|       20         |   ~8 min 30 sec               |      ~8 min 30 sec                    | exceeded threshold of 250ms (134012ms out of 210847ms) and 15% (64%) | Event loop utilization for /api/detection_engine/rules/_import exceeded threshold of 250ms (158122ms out of 516224ms) and 15% (31%)  |
|        5          |   ~34 min 16 sec              |     ~34 min 30 sec                 | - |  exceeded threshold of 250ms (314475ms out of 2066006ms) and 15% (15%) |

maximpn committed:
add performance measurement

banderror commented:
@maximpn Thank you for comparing performance in such a detailed comment explaining the results. I do agree with this:

> I suggest to classify this PR as rule import refactoring for better performance (APM) visibility.

Let's put this PR and the bug on hold and get back to it later in the 8.14 cycle.
